File: engine/app/parsing/parse.py
"""Dispatch a file to the right parser by extension.

Images are handled by the TypeScript server (Claude vision OCR) and never
reach the engine; scanned PDFs are detected here and bounced back for OCR.
"""

from pathlib import PurePosixPath, PureWindowsPath

from ..schemas import Segment
from .docx import parse_docx
from .pdf import parse_pdf
from .sheet import parse_workbook
from .types import ParsedDocument, ParseError, ParseOutcome

TEXT_EXTENSIONS = {".txt", ".md", ".csv"}
SUPPORTED_EXTENSIONS = {".pdf", ".docx", ".xlsx", ".xls"} | TEXT_EXTENSIONS


def _extension(name: str) -> str:
    # Handle both path styles defensively; uploads normally send a bare name.
    bare = PureWindowsPath(PurePosixPath(name).name).name
    return PurePosixPath(bare).suffix.lower()


def parse_file(name: str, data: bytes) -> ParseOutcome:
    ext = _extension(name)
    if not data:
        raise ParseError("File is empty.")

    if ext == ".pdf":
        return parse_pdf(data)
    if ext == ".docx":
        return parse_docx(data)
    if ext in (".xlsx", ".xls"):
        return parse_workbook(data, ext)
    if ext in TEXT_EXTENSIONS:
        text = data.decode("utf-8", errors="replace").strip()
        if not text:
            raise ParseError("File is empty.")
        return ParsedDocument(segments=[Segment(text=text, location=None)])

    raise ParseError(f"Unsupported file type: {ext or name}")

File: server/src/ingest/parse.ts
import path from "node:path";
import { isSupportedImageType, transcribeImage } from "../anthropic.js";
import { parseDocx } from "./docx.js";
import { parsePdf } from "./pdf.js";
import { parseWorkbook } from "./sheet.js";
import type { ParsedDocument } from "./types.js";

export const ALLOWED_EXTENSIONS = [
  ".pdf", ".docx", ".txt", ".md", ".csv", ".xlsx", ".xls",
  ".png", ".jpg", ".jpeg", ".gif", ".webp",
] as const;

export function isAllowedFilename(name: string): boolean {
  return (ALLOWED_EXTENSIONS as readonly string[]).includes(path.extname(name).toLowerCase());
}

export async function parseFile(name: string, mime: string, data: Buffer): Promise<ParsedDocument> {
  const ext = path.extname(name).toLowerCase();

  switch (ext) {
    case ".pdf":
      return parsePdf(data);
    case ".docx":
      return parseDocx(data);
    case ".xlsx":
    case ".xls":
      return parseWorkbook(data);
    case ".txt":
    case ".md":
    case ".csv": {
      const text = data.toString("utf8").trim();
      if (!text) throw new Error("File is empty.");
      return { segments: [{ text, location: null }], pageCount: null };
    }
    case ".png":
    case ".jpg":
    case ".jpeg":
    case ".gif":
    case ".webp": {
      const mediaType = mime === "image/jpg" ? "image/jpeg" : mime;
      if (!isSupportedImageType(mediaType)) {
        throw new Error(`Unsupported image type: ${mime}`);
      }
      const text = await transcribeImage(data, mediaType);
      if (!text) throw new Error("No text or content could be extracted from this image.");
      return { segments: [{ text, location: "image content" }], pageCount: null };
    }
    default:
      throw new Error(`Unsupported file type: ${ext}`);
  }
}
